### PR TITLE
allow predicate functions to lstrip etc. and default to stripping unicode whitespace

### LINKED ISSUE #27211: unicode whitespace not recognized by lstrip()
Consider:

```
julia> a = " \U02009 foo"
"   foo"

julia> lstrip(a)
"  foo"
```

Note that `\U02009` is a `Unicode Character 'THIN SPACE'` character.

I don't know much about string processing or unicode, but it seems to me that `lstrip()` is doing something that's perhaps too naive:

```
const _default_delims = [' ','\t','\n','\v','\f','\r']
...
function lstrip(s::AbstractString, chars::Chars=_default_delims)
```

I realize you can override `chars` but I would suggest that we expand the default character set for string trimming to include unicode separators (ref https://www.fileformat.info/info/unicode/category/Zs/list.htm).

### PR DESCRIPTION
**Pull request by simonbyrne**
This is similar to #27232, except that the predicate function is the second argument. It does not require any deprecations.

Fixes #27211.

### CODE REVIEW

File: base/strings/util.jl
Comment (ararslan) [inline base/strings/util.jl]: The predicate function should always go first per our conventions, documented in the style guide in the manual.

**Review by ararslan — CHANGES_REQUESTED** [review]
I don't think we should deviate from our argument order conventions for this case.

File: base/strings/util.jl
Comment (KristofferC) [reply base/strings/util.jl]: It is a "general rule" not a "must". Having the function argument second here seems much more natural to me.

Comment (fredrikekre) [reply base/strings/util.jl]: guidelines

Comment (ararslan) [reply base/strings/util.jl]: I disagree wholeheartedly; it seems very unnatural to me.

**Comment by StefanKarpinski** [comment]
I agree with @ararslan here: predicate function as the first argument, character sets as the second; we can also allow a predicate as either the first or second argument as the user prefers since a string argument isn't callable.

**Comment by ararslan** [comment]
That should probably be fixed.

**Comment by Keno** [comment]
I'm also with @ararslan. It doesn't really bother me that the predicate and the chars are in different places in the argument list.

**Comment by JeffBezanson** [comment]
I'm also OK with adding a predicate-first-argument version and just leaving the others.

**Comment by StefanKarpinski** [comment]
@simonbyrne, your PR, your call. Just do whatever you see fit and merge it once CI passes.

**Comment by ararslan** [comment]
Can I make one last plea to put the function argument first?

**Comment by simonbyrne** [comment]
@ararslan I am sympathetic, and I have thought a lot about it, but my opinion is that:
1. keeping argument order somewhat consistent is preferable to adhering to the function first rule
2. if we really want to allow predicate function _also_ as a first argument, we can add that later without breakage.
3. it is too late to change the argument order altogether, but it is possible to do that via deprecation without silent breakage (since we don't allow `chars` to be a string).

So my call is to merge this PR.

**Comment by simonbyrne** [comment]
Also, I got rid of the other uses of `_default_delimiters`.

**Comment by ararslan** [comment]
> keeping argument order somewhat consistent is preferable to adhering to the function first rule

I very adamantly disagree with this, since the argument order is then _not_ consistent. This will be a case of "functions always come first, except for this one weird case." We had that with `finalizer` and that was subsequently changed.

> if we really want to allow predicate function also as a first argument, we can add that later without breakage

I can submit a PR to add it once this is merged, but it really makes more sense to me to just have `lstrip(f, s)` and `lstrip(s, c)` rather than `lstrip(f, s)`, `lstrip(s, c)`, and `lstrip(s, f)`. But if that's unacceptable I'll just add `lstrip(f, s)`.

I do appreciate your work here and that you've given a lot of thought to this--I don't want you to think otherwise.

**Comment by ararslan** [comment]
PR to fix the argument order:  #27605

**Pull request merged**
